lib/module/src/module.rs

```diff
@@ -86,6 +86,26 @@ pub struct FunctionDeclaration {
     pub signature: ir::Signature,
 }
 
+/// Error messages for all `Module` and `Backend` methods
+#[derive(Fail, Debug)]
+pub enum ModuleError {
+    /// Indicates an identifier was used before it was declared
+    #[fail(display = "Undeclared identifier {}", _0)]
+    Undeclared(String),
+    /// Indicates an identifier was used contrary to the way it was declared
+    #[fail(display = "Incompatible declaration of identifier {}", _0)]
+    IncompatibleDeclaration(String),
+    /// Indicates that an identifier was already declared
+    #[fail(display = "Duplicate declaration of identifier {}", _0)]
+    DuplicateDeclaration(String),
```

Comment:
Right now, duplicate declarations are fine as long as they're compatible. Is there a use for the `DuplicateDeclaration` error code?